### DISCUSSION TITLE
Generate batch/multicall for contract reads

### DISCUSSION THREAD

**Discussion by naps62** [wevm/wagmi#1897] [Ideas]
I often need to use multicall for existing contract reads. I haven't found a way to do that with the existing hooks. Perhaps it could be added to the generator for all view methods that receive arguments (maybe opt-in if you're concerned with polluting the output too much)?

```typescript
export function useNFTFullInfoReads(
  configs: Array<
    Omit<
      UseContractReadConfig<typeof nftABI, "fullInfo">,
      "abi" | "functionName"
    >
  > = [] as any
) {
  const { chain } = useNetwork();

  const reads = configs.map((config) => {
    const chainId = config.chainId ?? chain?.id;
    return {
      abi: nftABI,
      address: nftAddress[chainId as keyof typeof nftAddress],
      functionName: "fullInfo",
      ...config,
    };
  });

  return useContractReads({ contracts: reads });
}
```

**Comment by xenoliss** [comment]
Hey there, not totally facing the same need but I have a pretty similar code that is building the `reads` and then calling `useContractReads({ contracts: reads })`. 

The problem with that is we loose the typing and the method is returning `UseQueryResult<unknown[], Error>`. Any idea how to keep typing right ? Thanks !

**Reply by azevedo** [reply]
Assuming you're creating `reads` with a `map`, you can return each `read` as a literal type using `as const`.

```ts
const reads = accounts.map(
  (account) =>
      ({
        address: contractAddress,
        abi: contractABI,
        functionName: "balance",
        args: [account],
      } as const) // <----
);

const results = useContractReads({ contracts: reads });
//     ^^^^^  This array is properly typed
```

**Reply by clumsier** [reply]
Also encountering this issue and can't force it with TS fully.

Do we have maybe some solution to cast types and have `functionName` and `args` having Typescript support inside `map` function and also keeping `results` typed properly?

**Reply by hungdoansy** [reply]
> Also encountering this issue and can't force it with TS fully.
> 
> Do we have maybe some solution to cast types and have `functionName` and `args` having Typescript support inside `map` function and also keeping `results` typed properly?

Did you have a solution?
